avocado/core/runner.py

```diff
@@ -425,9 +428,22 @@ def sigtstp_handler(signum, frame):     # pylint: disable=W0613
         abort_reason = None
         result_dispatcher = self.job._result_events_dispatcher
 
+        previous_phase = test_status.status.get('phase', 'UNKNOWN')
         while True:
+            phase = test_status.status.get('phase', 'UNKNOWN')
             try:
-                if time.time() >= deadline:
+                now = time.time()
+                if phase == 'SETUP' and setup_timeout:
+                    deadline = setup_deadline
+                elif phase != 'SETUP' and previous_phase == 'SETUP':
+                    # setup, which has its own timeout has finished,
+                    # so reset test deadline
+                    test_deadline = now + timeout
+                    if job_deadline is not None and job_deadline > 0:
+                        deadline = min(test_deadline, job_deadline)
+                    else:
+                        deadline = test_deadline
+                if now >= deadline:
                     abort_reason = "Timeout reached"
```

Comment:
Consider to change abort_reason to track it occurred on setup or test phase.

Reply:
Yep, good point.